src/node.cc

```diff
@@ -2498,36 +2502,96 @@ struct node_module* get_linked_module(const char* name) {
   return mp;
 }
 
-// DLOpen is process.dlopen(module, filename).
+struct DLib {
+  std::string filename_;
+  std::string errmsg_;
+  void* handle_;
+  int flags_;
+
+  void* Handle() {
+    return handle_;
+  }
+
+  const char* GetError() {
+    return errmsg_.c_str();
+  }
+
+#ifdef __POSIX__
+
+  static const int kDefaultFlags = RTLD_LAZY;
+
+  bool Open() {
+    handle_ = dlopen(filename_.c_str(), flags_);
+    if (handle_ != nullptr)
+      return true;
+    errmsg_ = dlerror();
+    return false;
+  }
+
+  void Close() {
+    if (handle_ != nullptr)
+      dlclose(handle_);
+  }
+
+#else  // !__POSIX__
+
+  static const int kDefaultFlags = 0;
+  uv_lib_t lib_;
+
+  bool Open() {
+    int ret = uv_dlopen(filename_.c_str(), &lib_);
+    if (ret == 0) {
+      handle_ = static_cast<void*>(lib_.handle);
+      return true;
+    }
+    errmsg_ = uv_dlerror(&lib_);
+    uv_dlclose(&lib_);
+    return false;
+  }
+
+  void Close() {
+    uv_dlclose(&lib_);
+  }
+#endif
+};
```

Comment:
Isn't calling `Close` necessary in the destructor? I see that we are explicitly calling `Close` everywhere, but I thought cleaning up in destructor is the best practice.

Reply:
Is this a blocker for landing?

Reply:
The handle should only be closed on error.  It needs to stay open on success, otherwise the shared object would be unloaded again.

Reply:
We are not closing everywhere. The handle is only closed on error. See https://github.com/nodejs/node/pull/12794#discussion_r137767595